xarray/core/computation.py

```diff
@@ -1382,6 +1382,182 @@ def _cov_corr(da_a, da_b, dim=None, ddof=0, method=None):
         return corr
 
 
+def cross(a, b, dim):
```

Comment:
We could add types here, but not required

Reply:
I haven't really used types before. But this is what I had in mind:
```python
def cross(
    a: Union[DataArray, Variable], b: Union[DataArray, Variable], dim: Hashable
) -> Union[DataArray, Variable]:
```
So `DataArray` has to be imported for this to work, right? But that would create cyclic dependencies. How is that usually worked around?

Reply:
I gave it a go with lots of errors as a result. To fix those it seems like I have to mess with other files as well, and mypy seems to fail on code paths that can't happen with those types which I found rather annoying. I think I'll skip this unless you have some code suggestions that fixes all these errors?

Reply:
Yeah, I had to do something pretty elaborate to persuade mypy that it's a `DataArray`. Fine to leave it given the additional code complication — your call.

For posterity:

```diff
diff --git a/xarray/core/computation.py b/xarray/core/computation.py
index 7bcd810d..038d6a3c 100644
--- a/xarray/core/computation.py
+++ b/xarray/core/computation.py
@@ -36,6 +36,7 @@
 if TYPE_CHECKING:
     from .coordinates import Coordinates  # noqa
     from .dataset import Dataset
+    from .dataarray import DataArray
 
 _NO_FILL_VALUE = utils.ReprObject("<no-fill-value>")
 _DEFAULT_NAME = utils.ReprObject("<default-name>")
@@ -1382,7 +1383,9 @@ def _cov_corr(da_a, da_b, dim=None, ddof=0, method=None):
         return corr
 
 
-def cross(a, b, dim):
+def cross(
+    a: Union["DataArray", "Variable"], b: Union["DataArray", "Variable"], dim: Hashable
+) -> Union["DataArray", "Variable"]:
     """
     Return the cross product of two (arrays of) vectors.
 
@@ -1526,11 +1529,17 @@ def cross(a, b, dim):
         # Arrays have different sizes. Append zeros where the smaller
         # array is missing a value, zeros will not affect np.cross:
         i = 1 if arrays[0].sizes[dim] > arrays[1].sizes[dim] else 0
+        array_i, array_j = (
+            (arrays[1], arrays[0])
+            if arrays[0].sizes[dim] > arrays[1].sizes[dim]
+            else (arrays[0], arrays[1])
+        )
 
-        if all([getattr(arr, "coords", False) for arr in arrays]):
+        # if all([isinstance(arr, DataArray) for arr in arrays]):
+        if isinstance(array_i, DataArray) and isinstance(array_j, DataArray):
             # If the arrays have coords we know which indexes to fill
             # with zeros:
-            arrays[i] = arrays[i].reindex_like(arrays[1 - i], fill_value=0)
+            arrays[i] = array_i.reindex_like(array_j, fill_value=0)
         elif arrays[i].sizes[dim] == 2:
             # If the array doesn't have coords we can can only infer
             # that it is composite values if the size is 2:

```

I then got a numpy type error, which I think we can ignore.

And notably we actually want this function to be generic over `DataArray` and `Variable` — the current construction doesn't map the return value to the argument types.